pkg/scheduler/admission/podnodeconstraints/admission.go

```diff
@@ -0,0 +1,163 @@
+)
+
+func init() {
+	admission.RegisterPlugin("PodNodeConstraints", func(c clientset.Interface, config io.Reader) (admission.Interface, error) {
+		pluginConfig, err := readConfig(config)
+		if err != nil {
+			return nil, err
+		}
+		return NewPodNodeConstraints(pluginConfig), nil
+	})
+}
+
+func NewPodNodeConstraints(config *api.PodNodeConstraintsConfig) admission.Interface {
+	return &podNodeConstraints{
+		config:  config,
+		Handler: admission.NewHandler(admission.Create),
+	}
+}
+
+type podNodeConstraints struct {
+	*admission.Handler
+	client client.Interface
+	config *api.PodNodeConstraintsConfig
+}
+
+var _ = oadmission.WantsOpenshiftClient(&podNodeConstraints{})
+var _ = oadmission.Validator(&podNodeConstraints{})
+
+func readConfig(reader io.Reader) (*api.PodNodeConstraintsConfig, error) {
+	if reader == nil || reflect.ValueOf(reader).IsNil() {
+		return &api.PodNodeConstraintsConfig{}, nil
+	}
+
+	obj, err := configlatest.ReadYAML(reader)
+	if err != nil {
+		return nil, err
+	}
+	if obj == nil {
+		return nil, nil
+	}
+	config, ok := obj.(*api.PodNodeConstraintsConfig)
+	if !ok {
+		return nil, fmt.Errorf("unexpected config object: %#v", obj)
+	}
+	// No validation needed since config is just 1 bool
+	return config, nil
+}
+
+func (o *podNodeConstraints) Admit(attr admission.Attributes) error {
+	switch {
+	// we want CREATE on primary resources we care about
+	case o.config == nil,
+		attr.GetSubresource() != "",
+		attr.GetOperation() != admission.Create:
+		return nil
+	}
+	switch attr.GetResource() {
+	case kapi.Resource("pods"),
+		kapi.Resource("replicationcontrollers"),
+		extensions.Resource("daemonsets"),
+		extensions.Resource("deployments"),
+		extensions.Resource("replicasets"),
+		extensions.Resource("jobs"),
+		deployapi.Resource("deploymentconfigs"):
+		// These are the only types we care about
+	default:
+		return nil
+	}
+	ps, err := o.getPodSpec(attr)
+	if err == nil {
+		return o.admitPodSpec(attr, ps)
+	}
+	return nil
+}
+
+// extract the PodSpec from the pod templates for each object we care about
+func (o *podNodeConstraints) getPodSpec(attr admission.Attributes) (kapi.PodSpec, error) {
+	switch r := attr.GetObject().(type) {
+	case *kapi.Pod:
+		return r.Spec, nil
+	case *kapi.ReplicationController:
+		return r.Spec.Template.Spec, nil
+	case *extensions.DaemonSet:
+		return r.Spec.Template.Spec, nil
+	case *extensions.Deployment:
+		return r.Spec.Template.Spec, nil
+	case *extensions.ReplicaSet:
+		return r.Spec.Template.Spec, nil
+	case *extensions.Job:
+		return r.Spec.Template.Spec, nil
+	case *deployapi.DeploymentConfig:
+		return r.Spec.Template.Spec, nil
+	}
+	return kapi.PodSpec{}, kapierrors.NewInternalError(fmt.Errorf("No PodSpec available for supplied admission attribute"))
+}
+
+// validate PodSpec if NodeName or NodeSelector are specified
+func (o *podNodeConstraints) admitPodSpec(attr admission.Attributes, ps kapi.PodSpec) error {
+	if len(ps.NodeName) > 0 || len(ps.NodeSelector) > 0 {
```

Comment:
I would think that it should be the same permission check that gets applied. The objective of the admission controller is to prevent users from targeting specific nodes for their pods. In this sense, specifying the NodName and the NodeSelector can have the same effect. 

At some point, the restriction on NodeName and NodeSelector could be applied individually - you could restrict either one or both. This seems to have been lost in various discussions/decisions/refactors. Since this flexibility is good-to-have for other use cases and doesn't hurt, I have asked @dobbymoodge to make that change.